### ISSUE TITLE
prefer-optional rule is incorrect

### ISSUE THREAD

**Issue by mad-gooze** [SonarSource/SonarTS#871]
**I want to provide the following feedback.**

**Rule key:** prefer-optional

**Example**
prefer-optional requires this declaration
```typescript
interface Person1 {
    name: string | undefined;
}
```
to be rewritten to 
```typescript
interface Person2 {
    name?: string;
}
```

But types `Person1` and `Person2` are different:
```typescript
const person2: Person2 = {}; // this is correct
const person1: Person1 = {}; // this is incorrect, name is missing but required
```

So `Person1` and `Person2` are semantically different.

**Comment by saberduck** [comment]
@mad-gooze thanks for feedback, I think you are right and we should drop this rule. cc @vilchik-elena

**Comment by atsu85** [comment]
in additon to what @mad-gooze pointed out, you may have types where some properties are optional and others, that are required, but can accept `undefined` value
```TypeScript
interface Person1 {
    name: string | undefined; // property must not be omitted (even tough value can be undefined)
    nickName?: string; // property can be omitted
}
```

**Comment by vilchik-elena** [comment]
Hi @atsu85 and @mad-gooze, 

I reworked RSPEC to report only on redundant usages: https://jira.sonarsource.com/browse/RSPEC-4782

WDYT? does it make sense to you like this?

**Comment by atsu85** [comment]
@vilchik-elena, would the rule complain if optional property has type that includes `| undefined`? For example:
```TypeScript
type StringOrUndefined = string | undefined;

interface Person {
  address? : StringOrUndefined;   // I think it shouldn't complain
}
```

**Comment by vilchik-elena** [comment]
> @vilchik-elena, would the rule complain if optional property has type that includes | undefined?

It depends on how we implement it:) but thanks for the point

**Comment by vilchik-elena** [comment]
> I'd propose changing the error message. Perhaps smth like this:
>Either ? or | undefined should be removed (depending what makes more sense in this specific use-case)

In fact that's the case already, the message will be `Consider removing 'undefined' type or '?' specifier, one of them is redundant.` (see "Message" field of the ticket). The text in comment is just to explain this particular case, and there are 2 different cases to show that it depends on property and context.

**Comment by atsu85** [comment]
Nice, i didn't notice it initially
